File: 416-PartitionEqualSubsetSum/416-PartitionEqualSubsetSum.py
# Last updated: 6/11/2026, 11:33:20 AM
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        
        totsum = sum(nums)
        

        if(totsum%2!=0):
            return False

        target = totsum//2

        dp = [[-1]*(10000) for _ in range(201)]
        
        def rec(ind, sumr):
            if(ind==len(nums)):
                if(sumr==target):
                    return True
                return False

            if(sumr>target):
                return False
            
            if(dp[ind][sumr]!=-1):
                return dp[ind][sumr]
            

            take = rec(ind+1, sumr+nums[ind])
            not_take = rec(ind+1, sumr)

            dp[ind][sumr] = take or not_take
            return dp[ind][sumr]

            
        return rec(0, 0)

File: 416-PartitionEqualSubsetSum/416-PartitionEqualSubsetSum.cpp
// Last updated: 6/11/2026, 11:33:13 AM
class Solution {
public:
    int dp[201][20001];
    bool helper(int ind, int sl, vector<int> &nums){
        if(sl<0) return 0;
        
        if(ind == nums.size()){
            return (sl == 0);
        }
        if(dp[ind][sl]!=-1) return dp[ind][sl];

        return dp[ind][sl] = helper(ind+1, sl-nums[ind], nums) || helper(ind+1, sl, nums);
    }


    bool canPartition(vector<int>& nums) {
        int tsum = accumulate(nums.begin(), nums.end(), 0);
        if(tsum%2!=0){
            return 0;
        }

        int hsum = tsum/2;
        for(int i=0; i<=nums.size(); i++){
            for(int j=0; j<=hsum; j++){
                dp[i][j] = -1;
            }
        }

        return helper(0, hsum, nums);

    }
};